autogen/agentchat/conversable_agent.py

```diff
@@ -719,14 +719,20 @@ def reset_consecutive_auto_reply_counter(self, sender: Optional[Agent] = None):
         else:
             self._consecutive_auto_reply_counter[sender] = 0
 
-    def clear_history(self, agent: Optional[Agent] = None):
+    def clear_history(self, agent: Optional[Agent] = None, preserve_messages: Optional[int] = None):
         """Clear the chat history of the agent.
 
         Args:
             agent: the agent with whom the chat history to clear. If None, clear the chat history with all agents.
+            preserve_messages: the number of newest messages to preserve in the chat history.
         """
         if agent is None:
-            self._oai_messages.clear()
+            if preserve_messages:
+                for key in self._oai_messages:
+                    # Remove messages from history except last `preserve_messages` messages.
+                    self._oai_messages[key] = self._oai_messages[key][-preserve_messages:]
+            else:
+                self._oai_messages.clear()
         else:
             self._oai_messages[agent].clear()
```

Comment:
Shall we log a warning that when `agent` is provided , `preserve_messages` is ignored here? Alternatively, I personally would raise an ValueError to be more defensive against cases where one parameter is ignored.

Reply:
@joshkyh Main idea when both of arguments (agent name and nr of last messages to preserve) are provided is to clear history of messages for that concrete agent except of last X messages. History for other agents is leaven untouched.

Thanks for suggestion about logging, I also thought that it worth to print info if preserve_messages provided, and wouldn't implement it previousely to not make code too long. But as you also think it's a good idea to log preserve_messages to user, I implemented that feature in the last commit.

Can you please explain what do you mean about raising the Exception/Warning, in which situation it should be raised? I didn't undertood the point.

Reply:
Hi @GregorD1A1,

Apologies, I realised I wasn't clear too. Points [1], [2], and [3] in my previous message were all options related to the issue when both `agent` and `preserve_messages` are supplied. I think we are picking [3], which is ok. So, we don't have to discuss [1] logging a warning nor [2] raising an error.

I think groupchat.py L478 - 490 is clear.

However, can you help me understand -- In conversable_agent.py https://github.com/microsoft/autogen/blob/62ca13edd57d5777c9bafb68bdab6d06d131e20f/autogen/agentchat/conversable_agent.py#L737 L737 meant that `agent` is not None. Why didn't we need two cases like L730 and L734 to differentiate between when `preserve_messages` is int vs. is None? It seems like at L737 we are clearing all history of `agent` regardless of `preserve_messages`. Did I get it right?

Reply:
Ok, no problem ;)

So, let's make it clear. Functionality from L737       self._oai_messages[agent].clear() to clear history with between agent and another agent was here before. To be honest, I don't understand why do we need to be able to clean history with concrete agent - it seems to be used only in L624             self.clear_history(recipient), in the _prepare_chat function. Here we are providing recepient history with whom needs to be cleared. I don't understand why here we can't just clear all the history for the agent without providing recipient - if we do so, we'll be able to simplify code. @sonichi maybe you know the reason?

Anyway, 'agent' argument name in 'clear_history' function seems to have ambigous name, as we have agent which history we clearing and agent with whom history we clearing. Maybe better to change argument name to 'agent_to_clear_history_with' or 'agent_partner' or something like that?

Answering your question @joshkyh, when user clears history manually (that's PR case), he never entered L737 as there will never argumet 'agent' provided to 'clear_history' function, so that's why any additional differentiation was provided. Hope I understend good and answered all your questions?

Reply:
> Answering your question @joshkyh, when user clears history manually (that's PR case), he never entered L737 as there will never argumet 'agent' provided to 'clear_history' function, so that's why any additional differentiation was provided. Hope I understend good and answered all your questions?

Yeap thanks, you've answered my question. It's not used in the codebase, and unlikely to be used by the user. I'll mark as resolve then.

Reply:
Thanks for suggestion, I agree it could be latent bug here. Tried to implement code solution as you suggested in [3], but realized that I can't find test case to check if my code works or not (when providing recipient agent name to function, there history with concrete agent was clear at beginning), so decided to use warning message instead as you suggested in [1]. Also changed variable name from 'agent' to 'recipient' to avoid ambiguouty. Is it ok now?

Reply:
Looks good to me, thanks for taking the suggestions on-board and the contributions. :)